File: leetcode/sum-of-nodes-with-even-valued-grandparent.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumEvenGrandparent = function(root) {
    const result = [];
    traversal(root, result);
    
    return result.length === 0 ? 0 : result.reduce((a, b) => a + b);
};

function traversal(node, result) {
    if(node === null) return;
    if(node.val % 2 === 0) {
        traversalEdgeNode(node, result, 0);
    }
    
    traversal(node.left, result);
    traversal(node.right, result);
}

function traversalEdgeNode(node, result, step) {
    if(node === null) return;
    if(step === 2) return result.push(node.val);
    
    traversalEdgeNode(node.left, result, step + 1);
    traversalEdgeNode(node.right, result, step + 1);
}

File: leetcode/sum-of-nodes-with-even-valued-grandparent.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumEvenGrandparent(self, root: TreeNode) -> int:
        ans = []
        
        def traversal(node):
            if node is None:
                return
            
            if node.val % 2 == 0:
                traversal_edge_node(node, 0)
            
            traversal(node.left)
            traversal(node.right)
        
        def traversal_edge_node(node, step):
            if node is None: 
                return
            
            if step == 2:
                ans.append(node.val)
                return
            
            traversal_edge_node(node.left, step + 1)
            traversal_edge_node(node.right, step + 1)
            
        traversal(root)
        return reduce(lambda a, b: a + b, ans, 0) if ans else 0
